daemons/based/based_notify.c

```diff
@@ -132,7 +132,7 @@ attach_cib_generation(xmlNode *msg)
     xmlNode *generation = pcmk__xe_create(wrapper, PCMK__XE_GENERATION_TUPLE);
 
     if (the_cib != NULL) {
-        pcmk__xe_copy_attrs(generation, the_cib);
+        pcmk__xe_copy_attrs(generation, the_cib, false);
```

Comment:
That's true. This is also a good catch. `cibadmin.c` has the following in the help text:
```
           "Increase configuration version to prevent old configurations from "
               "being loaded accidentally:\n\n"
           "\t# cibadmin --modify --xml-text "
               "'<" PCMK_XE_CIB " " PCMK_XA_ADMIN_EPOCH
                   "=\"" PCMK_XA_ADMIN_EPOCH "++\"/>'\n\n"
```

Previously I was looking for `"++"` since looking at all `++` would've been infeasible.

I expected the `bump()` method to bump the admin epoch but in fact it bumps the epoch. Still, bumping the admin epoch seems like more of a use case for a method than for some obscure syntax.

Reply:
It would be for CIB updates, generally via cibadmin rather than C calls. A tool could build an update and put `++` in the admin_epoch.

Reply:
I'm leaning to keeping the functionality. It is useful, to avoid having to read and parse the CIB before submitting a simple change. And if we've been using it as an example all this time, it's more likely someone somewhere is using it. I'd be fine with deprecating `+=` but that's so trivial we may as well keep it too.

Reply:
The main reason I'd love to deprecate and drop it (unless I'm forgetting some other big reason) is the asymmetry. I don't think a replace operation should treat "value++" as a literal if a modify operation treats it as "value=<existing_value>+1".
* It makes behavior less consistent.
* It's harder to reason about, both due to the above inconsistency and because the XML attr value that we set is not the XML attr value that we get.
* It makes the code more difficult to maintain.

> It is useful, to avoid having to read and parse the CIB before submitting a simple change

Yep. IMO that doesn't outweigh the disadvantages of keeping it. Prior to this PR, it was used only once in internal code, and now zero times. It's not used on any performance-critical path. Having to do an extra query before incrementing the failcount during a simulation, or as a user before updating the admin_epoch, isn't a big deal. For the latter case, we could add a method.

> it's more likely someone somewhere is using it

Maybe. (Having worked support for 4.5 years, I wouldn't be surprised if no one is using that to be honest with you.) But that's why we'd deprecate it instead of dropping it immediately.

Reply:
> it's more likely someone somewhere is using it

also on this note, we never explicitly documented it AFAIK, only indirectly as part of that `admin_epoch` example. (that's a weak argument on its own but it's worth noting)

Reply:
I'm OK with deprecating it, but let's take it out of the example, log a deprecation warning when it's used, and keep it for the 3.0 series

It's not a big loss for interactive usage, but for automation it could be a serious pain. Consider an ansible role that modifies the CIB and bumps admin_epoch using `++`. It might be worth keeping support for just admin_epoch.

Reply:
Another option would be to have a "hidden" attribute, something like `__increment__="true"` or `pcmk__increment="true"`, that `cibadmin` can set. If true, expand anything like `value="value++"` in the input XML. Otherwise, treat all values literally. We don't necessarily have to publicly document it, although we could. (We can ignore `+=`.) It could apply to both replace and modify, whereas the current behavior only works for modify.

That way, the default behavior is intuitive, but there's still a way to request an atomic increment.

I'm reluctant to add another obscure attribute but it could give the best of both worlds.

Reply:
> I'm reluctant to add another obscure attribute but it could give the best of both worlds.

The only part this wouldn't resolve is "this sucks to maintain." But it would be comparable to what we have now, at least.

Reply:
> Another option would be to have a "hidden" attribute, something like `__increment__="true"` or `pcmk__increment="true"`, that `cibadmin` can set. If true, expand anything like `value="value++"` in the input XML. Otherwise, treat all values literally. We don't necessarily have to publicly document it, although we could. (We can ignore `+=`.) It could apply to both replace and modify, whereas the current behavior only works for modify.
> 
> That way, the default behavior is intuitive, but there's still a way to request an atomic increment.
> 
> I'm reluctant to add another obscure attribute but it could give the best of both worlds.

We wouldn't need a special attribute; we could add a CIB call option.

Reply:
Oh you're right.

I lean toward using simply "++" instead of "name++" as the value for the new way. Simplifies parsing. As for "+=", it would be about as easy to keep it as it is to drop it. I guess keep it. Otherwise if it turns out we need it, adding it would be a breaking change.

Reply:
Agreed, but we should keep `name++` so that any existing scripts don't break. If only cibadmin uses the new option, it makes everything else more efficient while preserving the one reasonable use case. The CIB digests prevent simple editing of the file, so everyone really does have to go through cibadmin. I would also use it in crm_attribute's update option, which can go directly through the CIB API. I don't think any other tools would benefit from it.

Reply:
> we should keep `name++` so that any existing scripts don't break

The way I envisioned it, existing scripts would break anyway at a compatibility break. I figured we'd add a CLI option for the increment behavior. That way the user can specify that they want `"++"`/`"+=X"` to expand to "existing value plus one/plus X". Otherwise, no attempt is made to expand values, and what you input is what you get.

(Suppose that for some strange reason a user wants to set `"++"` as the literal value; they should be able to do so without any surprises. If they want special behavior, they can set a flag requesting it.)

Until some compatibility break, also preserve the existing "expand `value++` to `<value + 1>` by default for modify ops" behavior. But replace the example in the help output, with the new syntax that includes the CLI flag.

Reply:
planning to demo it as soon as I'm able and you can see if you like it or not. Classes today so might not be until tomorrow

Reply:
I wouldn't bother with a new option or syntax, just have cibadmin support the current syntax

Reply:
That's easy but leaves us in the same boat as now, where modify and replace behave differently (making replace do it automatically would be a breaking change), and where certain values are treated as "special" and transformed without it being requested

It's gonna have to support the current syntax until a compat break regardless, I mean going forward though

Realistically, the impact is minimal whatever we decide. I'm only coming at this from a "principle of least astonishment" POV. To me the least astonishing scenario is that I can set whatever value I want and it's treated
* literally
* the same regardless of how I request it

Reply:
I'm mainly concerned with the possibility of existing users depending on it, whether for admin_epoch via cibadmin or a permanent node attribute via crm_attribute. Both uses are actually reasonable, so I think it's worth keeping support. I'm thinking:

Keep the current syntax, forever, because it's not a big deal and it's one less thing existing users would have to change.

Add two new CIB call options that both enable the behavior. One logs a deprecation warning when the syntax is used.

Add command-line options to cibadmin and crm_attribute to enable the behavior. Show the new option in the cibadmin example. If the option is specified, they should pass the new call option without the warning; if the option is not specified, they should pass the new call option with the warning. The end result is that the behavior is still always supported, but the deprecation warning is logged if the new command-line option is not used.

Drop support for the deprecated usage in 3.1.0 or later.

Reply:
This seems like a good approach

Reply:
It is a *little* bit awkward that we treat these updates as scores, since one of the main anticipated use cases is for bumping the `admin_epoch`. Which is definitely not a score. Other use cases may also be regular old integer counters. I don't want to open that can of worms though.